README.md
# remoteweatheraccess
Weather network software (client + server) for collecting data from remote weather stations.

This project contains a client and a server software and is written in Python. It currently supported weatherstations derived from the type TE923 (for example IROX PRO-X USB).

The programs are written in Python and require Python 3.3.

You will require the driver from that project:
http://te923.fukz.org/

Install it in the directory te923con.

Install a new Raspberry for a new station:

1. Clone the SD-card onto the new raspberry
2. Rename the computer in /etc/hostname and /etc/hosts
3. Set new WLAN-connection: set SSID and WLAN-key in /etc/network/interfaces
4. Replace the WiFi test IP in /usr/local/bin/testwifi.sh by the router address
5. Replace station file /opt/weatherstation/data/stationData.dat by the new one using python3.3 and the module "generate_stationfile.py" (owner + group: weatherstation)
6. Delete the last data storage file /opt/weatherstation/data/settings_###.dat (if existing)
7. Make sure that there are no data files EXP_##_##.CSV in the directory /opt/weatherstation/data/
8. Generate a new user on the FTP-server:
	htpasswd -p -b /etc/ftpd.passwd 'station name' $(openssl passwd -1 -noverify 'password')
	The warning can be ignored
	mkdir /srv/ftp/virtual/'station name'
	chown -R ftp:ftp /srv/ftp/virtual/'station_name'
9.	Adapt the file structure of the FTP-server:
	/srv/ftp/virtual/'station name'
		-> newData (ftp:ftp)
		-> newData/temp (weatherstation:weatherstation)
	chmod -w /srv/ftp/virtual/'station_name'	
	chmod +rwx /srv/ftp/virtual/'stationName'/newData
10. Generate internet page for the user
	mkdir  /srv/www/'station name'
	chown http-web:http-web /srv/www/'station name'
	cp /srv/www/ERL /srv/www/'station name'
	chown
	Adapt texts in /srv/www/'station name'/index.html
11.	Generate the data folder for the new station:
	mkdir /srv/weather/'station name'
	chown -R weatherstation:weatherstation /srv/weather/'station name'
12.	Add the station ftp-folder to incrod-control:
	incrontab -e -u weatherstation
	Add the line: /srv/ftp/virtual/'station name'/newData IN_CLOSE_WRITE python3 /opt/weatherstation/weatherserver.py $@/$#

-> The weather data can be obtained from http://station.com/'station name'


debian/changelog
remote-weather-access (0.2.0-1) xenial; urgency=medium

  * source package automatically created by stdeb 0.8.5

 -- Ralf Rettig <info@personalfme.de>  Sun, 14 Jan 2018 13:19:00 +0100


te923ToCSVreader.py
# RemoteWeatherAccess - Weather network connecting to remote stations
# Copyright(C) 2013-2016 Ralf Rettig (info@personalfme.de)
#
# This program is free software: you can redistribute it and / or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.If not, see <http://www.gnu.org/licenses/>


"""Reading weather data from TE923-stations.

Functions:
te923ToCSVreader:               Reads the weather data from a TE923-station and transfers it to a server using Wetterstation-compatible CSV-files.
reset_logger:                   Resets a logger instance.

Global variables:
log_file_name:                  Name of the log file where the status of the operations is stored (only relevant for Windows, under Linux the syslog is used)
station_type:                   Detailed name of the stations for which this script is working
sensor_list:                    List of all sensors read including the mapping to information in the Wetterstation CSV-files.
"""
from datetime import datetime as dt
from datetime import timedelta as timedelta
from collections import OrderedDict
import logging
import sys

import constants
import stationdata
import lastdata
import te923station
import pcwetterstation
import server

log_file_name = 'te923.log' # only relevant for Windows
station_type = 'TE923/924 (Mebus,Irox,Honeywell,Cresta)'
# Containing the list of all sensors read from the weather station and their mapping to the PC-Wetterstation format
# The tuples contain: ( index in import data, description of sensor, index in PC-Wetterstation software, data unit required by PC-Wetterstation )
# Irrelevant sensors indices are marked by 'none'
sensor_list = OrderedDict( [
    ('date',           (0,        'Datum',                                            '',     ''                  )),
    ('time',           ('none',   'Zeit',                                             '',     ''                  )),
    ('tempInside',     (1,        'Temp. I.',                                          1,     '\N{DEGREE SIGN}C'  )),
    ('humidInside',    (2,        'Feuchte I.',                                       17,     '%'                 )),
    ('tempOutside1',   (3,        'Temp. A. 1',                                        2,     '\N{DEGREE SIGN}C'  )),
    ('humidOutside1',  (4,        'Feuchte A. 1',                                     18,     '%'                 )),
    ('tempOutside2',   (5,        'Temp. A. 2',                                        3,     '\N{DEGREE SIGN}C'  )),
    ('humidOutside2',  (6,        'Feuchte A. 2',                                     19,     '%'                 )),
    ('tempOutside3',   (7,        'Temp. A. 3',                                        4,     '\N{DEGREE SIGN}C'  )),
    ('humidOutside3',  (8,        'Feuchte A. 3',                                     20,     '%'                 )),
    ('tempOutside4',   (9,        'Temp. A. 4',                                        5,     '\N{DEGREE SIGN}C'  )),
    ('humidOutside4',  (10,       'Feuchte A. 4',                                     21,     '%'                 )),
    ('tempOutside5',   (11,       'Temp. A. 5',                                        6,     '\N{DEGREE SIGN}C'  )),
    ('humidOutside5',  (12,       'Feuchte A. 5',                                     22,     '%'                 )),
    ('pressure',       (13,       'Luftdruck',                                       133,     'hPa'               )),
    ('UV',             (14,       'UV-X',                                              9,     'UV-I'              )),
    ('forecast',       (15,       'Vorhersage',                                   'none',     ''                  )),
    ('warning',        (16,       'Sturmwarnung',                                 'none',     ''                  )),
    ('windDir',        (17,       'Richtung',                                         36,     '\N{DEGREE SIGN}'   )),
    ('windSpeed',      (18,       'Wind',                                             35,     'km/h'              )),
    ('windGusts',      (19,       'Windb\N{LATIN SMALL LETTER O WITH DIAERESIS}en',   45,     'km/h'              )),
    ('tempWind',       (20,       'Temp. Wind',                                        8,     '\N{DEGREE SIGN}C'  )),
    ('rainCounter',    (21,       'Regen',                                            34,     'mm'                ))
    ] )

delta_t_tol = timedelta( minutes = 60 ); # All datasets read from the weather station with a timestamp being more than this tolerance time in the future are dismissed


def reset_logger(logger):
    """Resets a logger instance.
    
    Args:
    logger:                     Logger instance to be reset

    Returns:
    None

    Raises:
    None
    """
    if not sys.platform.startswith('linux'):
        # Close the logger
        logger.handlers[0].stream.close()
        logger.removeHandler(logger.handlers[0])


def te923ToCSVreader(data_folder, station_data_file_name, script_name):
    """Reads the weather data from a TE923-station and transfers it to a server using Wetterstation-compatible CSV-files.
    
    Args:
    data_folder:                Operation directory of the function. Here all status files and the temporary data files are stored. Under Windows also the log-file is stored here.
    station_data_file_name:     Name of the file that contains all information on the weather station.
    script_name:                Name of the calling script (used for logging purposes)

    Returns:
    None

    Raises:
    None
    """
    firstNewDataIndex = 0

    # Start the logger
    logger = logging.getLogger()
    if sys.platform.startswith('linux'):
        from logging.handlers import SysLogHandler
        syslog = SysLogHandler(address='/dev/log')
        formatter = logging.Formatter('%(name)s - %(app_name)s - %(levelname)s : %(message)s')
        syslog.setFormatter(formatter)
        logger.setLevel(logging.INFO)
        logger.addHandler(syslog)
        logger = logging.LoggerAdapter(logger, { 'app_name': script_name } )
    else:
        logging.basicConfig( filename=log_file_name, level=logging.INFO, format='%(asctime)s %(levelname)s: %(message)s', datefmt='%b %d %H:%M:%S' )

    try:
        # Read station data
        try:
            rain_calib_factor, station_name, station_height, storage_interval, ftp_passwd, ftp_server, port, ftp_folder = stationdata.read( data_folder + '/' + station_data_file_name )
        except Exception as e:
            raise RuntimeError( 'Station data file could not be read. Error description: %s.', repr(e) )
        settings_file_name = 'settings_' + station_name + '.dat'
        try:
            is_reading, last_read_dataset_time, last_read_dataset_raincounter = lastdata.read( data_folder + '/' + settings_file_name )
        except Exception:
            # In this case it is assumed that the program is executed for the first time and all datasets should be read
            last_read_dataset_time = dt.min
            is_reading = False

        # Check if the station currently reads
        if is_reading:
            raise RuntimeError( 'Another instance is already reading from the weatherstation TE923. Skipping reading.' )

        # Notify other processes via the settings file of the reading in progress
        lastdata.set_reading( data_folder + settings_file_name, True )

        curr_time = dt.now()
        trunc_curr_time = dt( curr_time.year, curr_time.month, curr_time.day, curr_time.hour, curr_time.minute )
        trunc_last_read_dataset_time = dt( last_read_dataset_time.year, last_read_dataset_time.month, last_read_dataset_time.day, last_read_dataset_time.hour, last_read_dataset_time.minute )
        try:
            if ( trunc_curr_time - trunc_last_read_dataset_time ) < 2 * timedelta( minutes = storage_interval ):
                # Read only the current data
                imported_data = te923station.readdata( False )

                # Replace the timestamp by the current system time (required because the weatherstation occasionally generates wrong timestamps)
                imported_data[0][ sensor_list[ 'date' ][ constants.import_index ] ] = int( dt.now().timestamp() )
            else:
                # Read all datasets from the weatherstation and after this the current dataset - as the reading of the full dataset usually takes longer than one storage cycle
                imported_data = te923station.readdata( True )
                latest_data = te923station.readdata( False )

                # Replace the timestamp by the current system time (required because the stations occasionally generate wrong timestamps)
                latest_data[0][ sensor_list[ 'date' ][ constants.import_index ] ] = int( dt.now().timestamp() )

                imported_data = imported_data + latest_data
        except Exception as e:
            raise RuntimeError( 'Error accessing weatherstation TE923. Error description: %s.', repr(e) )

        if ( len( imported_data ) == 0 ):
            raise RuntimeError( 'No data could be read from the weather station. Maybe no USB-connection is possible.' )
        else:
            # Reduce datasets to unsaved new datasets (it is assumed that the read data is sorted according to time)
            date_index = sensor_list[ 'date' ][ constants.import_index ]
            imported_data = [ x for x in imported_data if ( dt.fromtimestamp( int( x[ date_index ] ) ) > last_read_dataset_time ) ]

            # Ensure that the data contains no timestamps in the future
            curr_time = dt.now() # Compare to the current system time
            imported_data = [ x for x in imported_data if ( dt.fromtimestamp( int( x[ date_index ] ) ) < curr_time + delta_t_tol ) ];
        
            if ( len( imported_data ) > 0 ):         
                # If this is the first execution of the program or the last stored data is older than three storage steps, there is no other choice than using the rain counter value of the first dataset as reference
                new_first_read_dataset_time = dt.fromtimestamp( int( imported_data[ firstNewDataIndex ][ sensor_list[ 'date' ][ constants.import_index ] ] ) );
                if ( new_first_read_dataset_time - last_read_dataset_time ) > 3 * timedelta( minutes = storage_interval ):
                    last_read_dataset_raincounter = float( imported_data[ firstNewDataIndex ][ sensor_list[ 'rainCounter' ][ constants.import_index ] ] )
  
                # Write weather data to PC-Wetterstation CSV-files
                export_data, last_dataset_time, last_dataset_rain_counter = pcwetterstation.convertTo( imported_data, last_read_dataset_raincounter, sensor_list )
                new_data_file_list, num_new_datasets, first_time, last_time = pcwetterstation.write( data_folder, rain_calib_factor, station_name, station_height, station_type, export_data, list( export_data[0].keys() ), sensor_list )
                logger.info( 'Found %i new weather datasets from %s - %s', num_new_datasets, first_time.strftime('%d.%m.%Y %H:%M'), last_time.strftime('%d.%m.%Y %H:%M') )

                # Transfer all CSV-files to the server
                data_file_list = pcwetterstation.finddatafiles( data_folder )
                try:
                    transfered_file_name = server.transferto( ftp_server, station_name, ftp_passwd, ftp_folder, data_folder, data_file_list, port )
                    isSuccessfullTransfer = True;
                except Exception as e:
                    error_text = repr( e )
                    isSuccessfullTransfer = False;

                # Delete the CSV-files in any situation
                pcwetterstation.deletedatafiles( data_folder, data_file_list )

                # Store the next latest dataset only if the transfer to the server was successfull, otherwise there is a rollback
                if ( isSuccessfullTransfer ):
                    # The reading flag is also resetted here by default
                    lastdata.write( data_folder + '/' + settings_file_name, last_dataset_time, last_dataset_rain_counter )
                    logger.info( 'Weather data in the files %s (%s) was successfully transfered to FTP-server \'%s\' (user: \'%s\').', data_file_list, transfered_file_name, ftp_server, station_name )
                else:
                    raise RuntimeError( 'Weather data transfer to FTP-server \'%s\' (user: \'%s\') failed. Read weather data in the files %s (%s) was discarded. Error description: %s.', ftp_server, station_name, data_file_list, transfered_file_name, error_text )
            else:
                raise RuntimeError( 'No weather data found which was unprocessed.' )
    except Exception as e:
        lastdata.set_reading( data_folder + '/' + settings_file_name, False )
        logger.error( repr(e) )
        reset_logger(logger)
        sys.exit()

    reset_logger(logger)
